src/System.Security.Permissions/src/System/Security/Permissions/PrincipalPermission.cs

```diff
@@ -2,25 +2,471 @@
+                return;
+
+            // A demand passes when the grant satisfies all entries.
+            int count = m_array.Length;
+            bool foundMatch = false;
+            for (int i = 0; i < count; ++i)
+            {
+                // If the demand is authenticated, we need to check the identity and role
+                if (m_array[i].m_authenticated)
+                {
+                    IIdentity identity = principal.Identity;
+
+                    if ((identity.IsAuthenticated &&
+                         (m_array[i].m_id == null || string.Compare(identity.Name, m_array[i].m_id, StringComparison.OrdinalIgnoreCase) == 0)))
+                    {
+                        if (m_array[i].m_role == null)
+                        {
+                            foundMatch = true;
+                        }
+                        else
+                        {
+                            foundMatch = principal.IsInRole(m_array[i].m_role);
+                        }
+
+                        if (foundMatch)
+                            break;
+                    }
+                }
+                else
+                {
+                    foundMatch = true;
+                    break;
+                }
+            }
+
+            if (!foundMatch)
+                throw new SecurityException(SR.Security_PrincipalPermission);
+        }
+
+        public SecurityElement ToXml()
+        {
+            SecurityElement root = new SecurityElement("IPermission");
+
+            string typename = "System.Security.Permissions.PrincipalPermission";
+            root.AddAttribute("class", typename + ", " + GetType().Module.Assembly.FullName.Replace('\"', '\''));
+            root.AddAttribute("version", "1");
+
+            if (m_array != null)
+            {
+                int count = m_array.Length;
+                for (int i = 0; i < count; ++i)
+                {
+                    root.AddChild(m_array[i].ToXml());
+                }
+            }
+
+            return root;
+        }
+
+        public void FromXml(SecurityElement elem)
+        {
+            if (elem == null)
+                throw new ArgumentNullException(nameof(elem));
+
+            if (elem.Tag == null || !elem.Tag.Equals("Permission") && !elem.Tag.Equals("IPermission"))
+                throw new ArgumentException(SR.Argument_NotAPermissionElement);
+
+            string version = elem.Attribute("version");
+
+            if (version == null || (version != null && !version.Equals("1")))
+                throw new ArgumentException(SR.Argument_InvalidXMLBadVersion);
+
+            if (elem.InternalChildren != null && elem.InternalChildren.Count != 0)
+            {
+                int numChildren = elem.InternalChildren.Count;
+                int count = 0;
+
+                m_array = new IDRole[numChildren];
+
+                IEnumerator enumerator = elem.Children.GetEnumerator();
+
+                while (enumerator.MoveNext())
```

Comment:
Nit: `foreach`. Although, it might be cheaper to use `ArrayList`'s indexer with a for loop here, but then again, this probably isn't perf-critical, so foreach would be simplest.